op-service/eth/blob.go

```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	return opos
+}
+
+func (b *Blob) write(buffer []byte, opos int) int {
+	copy(b[opos:], buffer)
+	return opos + len(buffer)
+}
+
 // ToData decodes the blob into raw byte data. See FromData above for details on the encoding
 // format.
 func (b *Blob) ToData() (Data, error) {
-	data := make(Data, 4096*32)
-	for i := 0; i < 4096; i++ {
-		if b[i*32] != 0 {
-			return nil, fmt.Errorf("invalid blob, found non-zero high order byte %x of field element %d", b[i*32], i)
+	// check the version
+	if b[VersionOffset] != EncodingVersion {
+		return nil, fmt.Errorf("invalid blob, expected version %d, got %d", EncodingVersion, b[VersionOffset])
+	}
+
+	// decode the 3-byte big-endian length value into a 4-byte integer
+	outputLen := uint32(b[2])<<16 | uint32(b[3])<<8 | uint32(b[4])
+	if outputLen > MaxBlobDataSize {
+		return nil, fmt.Errorf("invalid blob, length prefix out of range: %d", outputLen)
+	}
+
+	// round 0 is special cased to copy only the remaining 27 bytes of the first field element into
+	// the output due to version/length encoding already occupying its first 5 bytes.
+	output := make(Data, MaxBlobDataSize)
+	copy(output[0:27], b[5:])
+
+	// now process remaining 3 field elements to complete round 0
+	opos := 28 // current position into output buffer
+	ipos := 32 // current position into the input blob
+	var err error
+	encodedByte := make([]byte, 4) // buffer for the 4 6-bit chunks
+	encodedByte[0] = b[0]
+	for i := 1; i < 4; i++ {
+		encodedByte[i] = b[ipos]
+		if opos, ipos, err = b.decodeFieldElement(opos, ipos, encodedByte, output); err != nil {
+			return nil, err
+		}
+	}
+	opos = reassembleBytes(opos, encodedByte, output)
+
+	// in each remaining round we decode 4 field elements (128 bytes) of the input into 127 bytes
+	// of output
+	for i := 1; i < Rounds && opos < int(outputLen); i++ {
+		for j := 0; j < 4; j++ {
+			// save the first byte of each field element for later re-assembly
+			encodedByte[j] = b[ipos]
+			if opos, ipos, err = b.decodeFieldElement(opos, ipos, encodedByte, output); err != nil {
+				return nil, err
+			}
 		}
-		copy(data[i*31:i*31+31], b[i*32+1:i*32+32])
+		opos = reassembleBytes(opos, encodedByte, output)
 	}
-	// extract the length prefix & trim the output accordingly
-	dataLen := binary.LittleEndian.Uint32(data[:4])
-	data = data[4:]
-	if dataLen > uint32(len(data)) {
-		return nil, fmt.Errorf("invalid blob, length prefix out of range: %d", dataLen)
+	output = output[:outputLen]
+	return output, nil
+}
+
+// decodeFieldElement decodes the next input field element by writing its lower 31 bytes into its
+// appropriate place in the output and checking the high order byte is valid.
+func (b *Blob) decodeFieldElement(opos, ipos int, encodedByte []byte, output []byte) (int, int, error) {
+	// two highest order bits of the first byte of each field element should always be 0
+	if b[ipos]&(0xC0) != 0 {
+		return 0, 0, fmt.Errorf("invalid blob, field element %d has a high order bit set", ipos)
```

Comment:
I do not see any test coverage of this error

Reply:
see first case of TestInvalidBlobDecoding